test/infrastructure/container/docker.go

```diff
@@ -121,55 +126,208 @@ func (d *docker) GetHostPort(ctx context.Context, name, portAndProtocol string)
+		close(outputErrors)
 	}()
 
 	select {
+	case err := <-inputErrors:
+		if err != nil {
+			return errors.Wrap(err, "error providing execution input")
+		}
+
 	case err := <-outputErrors:
 		if err != nil {
-			return fmt.Errorf("error reading output from container exec: %v", err)
+			return errors.Wrap(err, "error getting execution output")
 		}
 
 	case <-ctx.Done():
-		return err
+		return errors.Wrap(ctx.Err(), "operation cancelled")
+	}
+
+	retry := 0
+	for retry < 600 {
+		inspect, err := d.dockerClient.ContainerExecInspect(ctx, execID)
+		if err != nil {
+			return errors.Wrap(err, "failed to get exec status")
+		}
+
+		if !inspect.Running {
+			if status := inspect.ExitCode; status != 0 {
+				return errors.Errorf("exited with status: %d, %s", status, config.OutputBuffer)
+			}
+			break
+		}
+
+		time.Sleep(time.Millisecond * 500)
+		retry++
+	}
+
+	return nil
+}
+
+// ListContainers returns a list of all containers.
+func (d *docker) ListContainers(ctx context.Context, filters FilterBuilder) ([]Container, error) {
+	listOptions := types.ContainerListOptions{
+		All:     true,
+		Limit:   -1,
+		Filters: dockerFilters.NewArgs(),
+	}
+
+	// Construct our filtering options
+	for key, values := range filters {
+		for subkey, subvalues := range values {
+			for _, v := range subvalues {
+				if v == "" {
+					listOptions.Filters.Add(key, subkey)
+				} else {
+					listOptions.Filters.Add(key, fmt.Sprintf("%s=%s", subkey, v))
+				}
+			}
+		}
+	}
+
+	dockerContainers, err := d.dockerClient.ContainerList(ctx, listOptions)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list containers")
+	}
+
+	containers := []Container{}
+	for i := range dockerContainers {
+		container := dockerContainerToContainer(&dockerContainers[i])
+		containers = append(containers, container)
+	}
+
+	return containers, nil
+}
+
+// DeleteContainer will remove a container, forcing removal if still running.
+func (d *docker) DeleteContainer(ctx context.Context, containerName string) error {
+	return d.dockerClient.ContainerRemove(ctx, containerName, types.ContainerRemoveOptions{
+		Force:         true, // force the container to be delete now
+		RemoveVolumes: true, // delete volumes
+	})
+}
+
+// KillContainer will kill a running container with the specified signal.
+func (d *docker) KillContainer(ctx context.Context, containerName, signal string) error {
+	return d.dockerClient.ContainerKill(ctx, containerName, signal)
+}
+
+// GetContainerIPs inspects a container to get its IPv4 and IPv6 IP addresses.
+func (d *docker) GetContainerIPs(ctx context.Context, containerName string) (string, string, error) {
+	containerInfo, err := d.dockerClient.ContainerInspect(ctx, containerName)
+	if err != nil {
+		return "", "", errors.Wrap(err, "failed to get container details")
+	}
+
+	ip := ""
+	ip6 := ""
+	for _, net := range containerInfo.NetworkSettings.Networks {
+		ip = net.IPAddress
+		ip6 = net.GlobalIPv6Address
+		break
+	}
+
+	return ip, ip6, nil
```

Comment:
I don't think we should error here. I think it would make sense to only error if there was a problem retrieving the information about the container. The container not having an IP address wouldn't be an error in retrieval, so I think it would be up to the consuming code to check and decide whether not having an IP address is an error condition at that point.

In other words, the error state should be determined by the calling code that knows what it is trying to accomplish rather than the interface code that is just fetching data.

Reply:
You're right, sounds good to me!

Reply:
Adding a note to the godoc to point this out.